openshift-kube-apiserver/admission/storage/performantsecuritypolicy/admission.go

```diff
@@ -0,0 +1,166 @@
+	"k8s.io/client-go/informers"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/component-base/featuregate"
+	"k8s.io/klog/v2"
+	kapi "k8s.io/kubernetes/pkg/apis/core"
+)
+
+const (
+	// Plugin name
+	PluginName               = "storage.openshift.io/PerformantSecurityPolicy"
+	fsGroupChangePolicyLabel = "storage.openshift.io/fsgroup-change-policy"
+	selinuxChangePolicyLabel = "storage.openshift.io/selinux-change-policy"
+)
+
+var (
+	_ = initializer.WantsExternalKubeInformerFactory(&performantSecurityPolicy{})
+	_ = admission.MutationInterface(&performantSecurityPolicy{})
+	_ = initializer.WantsFeatures(&performantSecurityPolicy{})
+)
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName,
+		func(config io.Reader) (admission.Interface, error) {
+			return &performantSecurityPolicy{
+				Handler: admission.NewHandler(admission.Create),
+			}, nil
+		})
+}
+
+// performantSecurityPolicy checks and applies if a default FSGroupChangePolicy and SELinuxChangePolicy
+// should be applied to the pod.
+type performantSecurityPolicy struct {
+	*admission.Handler
+	storagePerformantSecurityPolicyFeatureEnabled bool
+	nsLister                                      corev1listers.NamespaceLister
+	nsListerSynced                                func() bool
+}
+
+// SetExternalKubeInformerFactory registers an informer
+func (c *performantSecurityPolicy) SetExternalKubeInformerFactory(kubeInformers informers.SharedInformerFactory) {
+	c.nsLister = kubeInformers.Core().V1().Namespaces().Lister()
+	c.nsListerSynced = kubeInformers.Core().V1().Namespaces().Informer().HasSynced
+	c.SetReadyFunc(func() bool {
+		return c.nsListerSynced()
+	})
+}
+
+func (c *performantSecurityPolicy) InspectFeatureGates(featureGates featuregate.FeatureGate) {
+	c.storagePerformantSecurityPolicyFeatureEnabled = featureGates.Enabled(featuregate.Feature(openshiftfeatures.FeatureGateStoragePerformantSecurityPolicy))
+}
+
+func (c *performantSecurityPolicy) ValidateInitialization() error {
+	if c.nsLister == nil {
+		return fmt.Errorf("%s plugin needs a namespace lister", PluginName)
+	}
+	if c.nsListerSynced == nil {
+		return fmt.Errorf("%s plugin needs a method to sync namespace lister", PluginName)
+	}
+	return nil
+}
+
+func (c *performantSecurityPolicy) Admit(ctx context.Context, attributes admission.Attributes, _ admission.ObjectInterfaces) error {
+	if !c.storagePerformantSecurityPolicyFeatureEnabled {
+		return nil
+	}
+
+	if attributes.GetResource().GroupResource() != kapi.Resource("pods") ||
+		len(attributes.GetSubresource()) > 0 {
+		return nil
+	}
+
+	pod, ok := attributes.GetObject().(*kapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attributes, fmt.Errorf("unexpected object: %#v", attributes.GetObject()))
+	}
+
+	currentFSGroupChangePolicy := extractCurrentFSGroupChangePolicy(pod)
+	ns, err := c.nsLister.Get(pod.Namespace)
+	if err != nil {
+		return fmt.Errorf("error listing pod namespace: %v", err)
```

Comment:
should this be `admission.NewNotFound` or `admission.NewForbidden` ?

Reply:
okay, why? Should we change all other openshift admission hooks just as the one I linked above? TBH - this seems rather arbitrary.

Reply:
The admission error funcs return apimachinery *StatusError, so they do affect what is returned to a client (status and body).

Reply:
I checked, and the "bare" error will be wrapped in a status so we are fine here.